src/scales/scale.time.js

```diff
@@ -300,7 +300,7 @@ function determineUnitForFormatting(scale, ticks, minUnit, min, max) {
 
 	for (i = ilen - 1; i >= UNITS.indexOf(minUnit); i--) {
 		unit = UNITS[i];
-		if (INTERVALS[unit].common && scale._adapter.diff(max, min, unit) >= ticks.length) {
+		if (INTERVALS[unit].common && scale._adapter.diff(max, min, unit) >= ticks.length - 1.1) { // extra 0.1 for epsilon
```

Comment:
Clearly, `- 1` is needed, but no need for epsilon if we use `moment.diff(time, unit)` in `adapter.diff()` instead of `moment.duration(moment.diff(time)).as(unit)`. 
```js
if (INTERVALS[unit].common && scale._adapter.diff(max, min, unit) >= ticks.length - 1) {
```
adapter.moment.js:
```js
diff: function(max, min, unit) {
    return moment(max).diff(moment(min), unit);
},
```

See https://momentjs.com/docs/#/displaying/difference/.
> `moment#diff` has some special handling for month and year diffs. It is optimized to ensure that two months with the same date are always a whole number apart.

Reply:
Ah, thanks! That's a much cleaner fix. Updated!